### ISSUE TITLE
Git revision is meaningless in certain environments

### ISSUE THREAD

**Issue by weaverba137** [ameisner/ci_reduce#21]
During GFA processing, the function `gfa_reduce.common.retrieve_git_rev()` is used to determine a hex string corresponding to the git revision of `gfa_reduce`. It is also used to try to get a git revision for `desimeter`. At least one of these hex strings is inserted into the GFA daily summary file.

At a low level, it is running `git rev-parse --short HEAD`.

However, in a typical NERSC environment, software is installed from a tagged version, so `git rev-parse` won't work at all.

What is the desired behavior for a tagged, installed version? It may be possible to retrieve the tag for desiInstalled packages at NERSC, but I would not want to try to make this function work under all circumstances in an arbitrary environment.

@schlafly, @sbailey?

**Comment by sbailey** [comment]
What I *wish* we could have for all products, but I don't know if it is technically doable without a lot of hassle:
`package.__githash__` would automatically contain the githash whether the package was formally installed or running out of a git clone.  "Automatically" is key here; I don't want a solution that requires humans to remember to run a command after every update, or even requires humans to remember to add some git hook to every repo they've cloned.  Never trust humans to do the right thing.

What *might* be doable with less hassle: have desiInstall add the `__githash__` to packages while installing tagged version, but it would *not* be in git branch "installations".  I'm not sure if that is easy or hard to add hooks like that for a pip install.  A utility function like `desiutil.git.get_hash(package_name)` would auto-detect if `__githash__` existed for a package and return that if so, or otherwise run `git rev-parse --short HEAD` and return that, including returning something meaningful like "unknown" if `__githash__` doesn't exist and it's not actually a git clone.

What probably is doable with minimal hassle: daily ops runs out of main for all packages so that we're always using the latest versions, and that could apply to gfa_reduce as well.  In that case, `git rev-parse --short HEAD` would work and would return the correct value.  Since gfa_reduce isn't updated often, we could also likely keep the latest tag in sync with the latest git hash in main anyway.

Detail: `git rev-parse --short HEAD` does *not* work for non-desi users running it in a desi-owned main installation, e.g.
```
[login14] whoami
sjbailey
[login14] pwd
/global/common/software/desi/perlmutter/desiconda/20230111-2.1.0/code/desispec/main
[login14 main] git rev-parse --short HEAD
fatal: detected dubious ownership in repository at '/global/common/software/desi/perlmutter/desiconda/20230111-2.1.0/code/desispec/main'
To add an exception for this directory, call:

	git config --global --add safe.directory /global/common/software/desi/perlmutter/desiconda/20230111-2.1.0/code/desispec/main
```

That provides the command to run to "fix" this (though I haven't tried it).  If we really want to go down this route as a standard thing to do, we should have desiInstall also run that for new main installs, and/or have wrapper code that can catch that exception e.g. if running gfa_reduce as a non-desi user while using pre-installed main packages.

**Comment by weaverba137** [comment]
At least some of that functionality is available via `setuptools_scm`.  I do know that it can set both `__githash__` and `__version__` but I'm more familiar with its operations on the latter.

In order to get that functionality, we would need to adhere even more strictly to modern Python package infrastructure standards, and possibly even do things like `pip install -e` (development mode) in order to trigger `setuptools_scm`.

I really don't want to dig into `desiInstall` for the purposes of completing the work PR #20. So I suggest just documenting that function for now and say it may lead to unexpected behavior when *not* run with desimodules = main and gfa_reduce = main.

**Comment by schlafly** [comment]
I'm not aware of anyone actually using this field in the gfa_reduce data, though it does seem nice to have.  From my perspective as far as gfa_reduce is concerned it would be good to have this but it should be low priority as long as the code is actually running.

**Comment by sbailey** [comment]
From Ben:
> I really don't want to dig into desiInstall for the purposes of completing the work PR [#20](https://github.com/desihub/gfa_reduce/pull/20). So I suggest just documenting that function for now and say it may lead to unexpected behavior when not run with desimodules = main and gfa_reduce = main.

Agreed

From Eddie:
> From my perspective as far as gfa_reduce is concerned it would be good to have this but it should be low priority as long as the code is actually running

Also agreed

**Comment by weaverba137** [comment]
OK, please take a look at the documentation: https://github.com/desihub/gfa_reduce/blob/d7b4c7f96974e0a2cfde5918f8e945f8c8bd7427/py/gfa_reduce/common.py#L250

**Comment by schlafly** [comment]
That documentation looks clear to me, thanks.

**Comment by weaverba137** [comment]
Closed by #20.
